components/gitpod-protocol/src/protocol.ts

```diff
@@ -968,9 +977,47 @@ export namespace AdditionalContentContext {
     }
 }
 
-export interface CommitContext extends WorkspaceContext, Commit {
+export interface CommitContext extends WorkspaceContext, GitCheckoutInfo {
     /** @deprecated Moved to .repository.cloneUrl, left here for backwards-compatibility for old workspace contextes in the DB */
     cloneUrl?: string
+
+    /**
+     * The clone and checkout information for additional repositories in case of multi-repo projects.
+     */
+    additionalRepositoryCheckoutInfo?: GitCheckoutInfo[];
+}
+
+export namespace CommitContext {
+    /**
+     * Creates an identifier pointing to the commits of the CommitContext and all sub-repo commit infos.
+     * The identifier is max 255 chars long.
+     * @param commitContext
+     * @returns identifier for set of commits
+     */
+    export function getCommitSHAs(commitContext: CommitContext): string {
+        if (!commitContext.additionalRepositoryCheckoutInfo || commitContext.additionalRepositoryCheckoutInfo.length === 0) {
+            return commitContext.revision;
+        }
+        // it's a multi-repo project, we need to store all commit shas and that needs to fit into 255 chars. Let's only store at most first seven.
+        const commits = [commitContext.revision, ... commitContext.additionalRepositoryCheckoutInfo.map(i => i.revision)];
+        const maxLength = Math.min(7, 255 / commits.length);
```

Comment:
Hm. It feels strange to have a hard limit here.

Why not use a hash function over all commits SHAs? For instance:
```
const hasher = createHash('sha256');
for (const revision of revisions) {
   hasher.update(revision);
}
return hasher.digest('hex');
```

Reply:
We need to know the actual commit shas later (for incremental prebuilds), so the string is unparsed again. Ideally, we would store commit objects, that consist of the repo and the commit information.

Reply:
Today I learned that `String.substring(int, float)` actually works in JavaScript. 🙈

Still, I'd feel more reassured without floating-point length variables:

```suggestion
        const maxLength = Math.min(7, Math.floor(255 / commits.length));
```

Reply:
I was able change this to the suggested hash, as I now use the context of the workspace to get the commits.